packages/libs/client/src/v2/client/client.ts

```diff
@@ -0,0 +1,245 @@
+   */
+  createSpv: (
+    requestKey: string,
+    targetChainId: string,
+    options?: INetworkOptions,
+  ) => Promise<string>;
+}
+
+interface IGetClient {
+  /**
+   * path the url of the host if you only use one host
+   */
+  (hostUrl: string): IClient;
+  /**
+   * path a function that let you decide about with host url you should be picked for each request based on networkId and chianId
+   * the default value returns kadena testnet or mainnet url based on networkId
+   */
+  (
+    hostAddressGenerator?: (networkId: string, chainId: string) => string,
+  ): IClient;
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+const getRequestStorage = (initial?: Map<string, string>) => {
+  const storage = new Map<string, string>(initial);
+
+  return {
+    add: (hostUrl: string, requestKeys: string[]) => {
+      const list = storage.get(hostUrl) ?? [];
+      requestKeys.forEach((requestKey) => {
+        storage.set(requestKey, hostUrl);
+      });
+    },
+    remove: (requestKeys: string[]) => {
+      requestKeys.forEach((requestKey) => {
+        storage.delete(requestKey);
+      });
+    },
+    groupByHost: () => {
+      const byHost = new Map<string, string[]>();
+      storage.forEach((url, requestKey) => {
+        const prev = byHost.get(url) ?? [];
+        byHost.set(url, [...prev, requestKey]);
+      });
+      return [...byHost.entries()];
+    },
+    getByHost: (hostUrl: string) => {
+      const list: string[] = [];
+      storage.forEach((url, requestKey) => {
+        if (url === hostUrl) {
+          list.push(requestKey);
+        }
+      });
+      return list;
+    },
+    get: (requestKey: string) => storage.get(requestKey),
+  };
+};
+
+export const getClient: IGetClient = (host = kadenaHostGenerator): IClient => {
+  const getHost = typeof host === 'string' ? () => host : host;
+  const storage = getRequestStorage();
+
+  function groupByHost(
+    requestKeys?: string[],
+    options?: IPollOptions & INetworkOptions,
+  ) {
+    let requestStorage = storage;
+    if (requestKeys !== undefined) {
+      const map = new Map<string, string>(
+        requestKeys.map((requestKey) => [
+          requestKey,
+          storage.get(requestKey) ??
+            getHost(options!.networkId, options!.chainId),
+        ]),
+      );
+      requestStorage = getRequestStorage(map);
+    }
+    return requestStorage.groupByHost();
+  }
+  function pollStatusFunction(
+    requestKeys?: string[] | string,
+    options?: IPollOptions & INetworkOptions,
+  ) {
+    const results = groupByHost(
+      typeof requestKeys === 'string' ? [requestKeys] : requestKeys,
+      options,
+    ).map(([hostUrl, requestKeys]) =>
+      pollStatus(hostUrl, requestKeys, options),
+    );
+
+    // merge all of the result in one object
+    const mergedResults = mergeAllPollRequestPromises(results);
```

Comment:
How does this work? I was kinda expecting a `await` before we actually get `mergedResults`? 

I was actually expecting something along the lines of `const mergedResults = await mergeAllPollRequestsPromises(results)`

Reply:
I will rename it to `mergedPollRequestPromises` actually we just return it here and let the user to decide await for that or await for each request separately